lsp-methods.el

```diff
@@ -852,10 +852,13 @@ directory."
 (defun lsp--suggest-project-root ()
   "Get project root."
   (or
-   (when (featurep 'projectile) (projectile-project-root))
+   (when (featurep 'projectile)
+     (projectile-ensure-project (projectile-project-root)))
    (when (featurep 'project)
-     (when-let ((project (project-current)))
-       (car (project-roots project))))))
+     (let ((project (project-current)))
+       (if project
+           (car (project-roots project))
+         default-directory)))))
```

Comment:
That means users will get nil when they do not have `project` installed? Is this the intended behavior?

Reply:
Yes. I don't like it, but @MaskRay wants this behavior. The first edition is my purpose, and I use it in my personal config.

Reply:
How about this?

```elisp
(defun lsp--suggest-project-root ()
  "Get project root."
  (or
   (when (featurep 'projectile)
     (projectile-ensure-project (projectile-project-root)))
   (when (featurep 'project)
     (when-let ((project (project-current)))
       (car (project-roots project))))
   default-directory))
```

Reply:
This one makes more sense to me. Not sure what are the advantages of the previous version.

Reply:
I am not very sure the background, but I think this version is safer and solid for all scenarios.
I updated and re-submitted the codes.

Reply:
I hope this function can return `nil` when the user does not want to start a language server.

Reply:
This is different from an outright blacklist:

*  jumping from a project file to `/usr/include/stdio.h` adds the new buffer to the session
* opening `/usr/include/stdio.h` directly does not start a language server

Reply:
@MaskRay now I understood what are your concerns with defaulting root and they look valid to me. I think at this point it is trade-off which "feature" we want to sacrifice. I will let @vibhavp decide whether to merge it. As a side note, the library handling is covered in #469 as allowing registration of library folder paths which are not going to be used for starting a lsp server.

Reply:
Can't agree any more. It's up to the requirement and needs a trade-off.
Anyway I use the workaround in my config for now.